routers/api/actions/artifacts.go

```diff
@@ -0,0 +1,505 @@
+	}
+
+	storagePath := fmt.Sprintf("tmp%d/%d-%d-%d.chunk", taskID, artifact.ID, start, end)
+
+	// use hashReader to avoid reading all body to md5 sum.
+	// it writes data to hasher after reading end
+	// if hash is not matched, delete the read-end result
+	r := &hashReader{
+		Reader: ctx.Req.Body,
+		Hasher: md5.New(),
+	}
+
+	// save chunk to storage
+	if _, err := ar.fs.Save(storagePath, r, -1); err != nil {
+		return -1, fmt.Errorf("save chunk to storage error: %v", err)
+	}
+
+	// check md5
+	if !r.Match(ctx.Req.Header.Get(artifactXActionsResultsMD5Header)) {
+		if err := ar.fs.Delete(storagePath); err != nil {
+			log.Error("Error deleting chunk: %s, %v", storagePath, err)
+		}
+		return -1, fmt.Errorf("md5 not match")
+	}
+
+	log.Debug("[artifact] save chunk %s, size: %d, artifact id: %d, start: %d, end: %d",
+		storagePath, contentSize, artifact.ID, start, end)
+
+	return length, nil
+}
+
+var invalidArtifactNameChars = []string{"\\", "/", "\"", ":", "<", ">", "|", "*", "?", "\r", "\n"}
+
+// checkArtifactName checks if the artifact name contains invalid characters.
+// If the name contains invalid characters, an error is returned.
+// The rules are from https://github.com/actions/toolkit/blob/main/packages/artifact/src/internal/path-and-artifact-name-validation.ts#L32
+func checkArtifactName(name string) error {
+	for _, c := range invalidArtifactNameChars {
+		if strings.Contains(name, c) {
+			return fmt.Errorf("artifact name contains invalid character %s", c)
+		}
+	}
+	return nil
+}
+
+func (ar artifactRoutes) uploadArtifact(ctx *context.Context) {
+	artifactID := ctx.ParamsInt64("artifactID")
+
+	artifact, err := actions.GetArtifactByID(ctx, artifactID)
+	if err != nil {
+		log.Error("Error getting artifact: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	// itemPath is generated from upload-artifact action
+	// it's formatted as {artifact_name}/{artfict_path_in_runner}
+	itemPath := util.PathJoinRel(ctx.Req.URL.Query().Get("itemPath"))
+	taskID := ctx.ParamsInt64("taskID")
+	artifactName := strings.Split(itemPath, "/")[0]
+	if err = checkArtifactName(artifactName); err != nil {
+		log.Error("Error checking artifact name: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	// get upload file size
+	fileSize, contentLength, err := ar.getUploadFileSize(ctx)
+	if err != nil {
+		log.Error("Error getting upload file size: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	// save chunk
+	chunkAllLength, err := ar.saveUploadChunk(ctx, artifact, contentLength, taskID)
+	if err != nil {
+		log.Error("Error saving upload chunk: %v", err)
+		ctx.Error(http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	// if artifact name is not set, update it
+	if artifact.ArtifactName == "" {
+		artifact.ArtifactName = artifactName
+		artifact.ArtifactPath = itemPath // path in container
```

Comment:
Looks like `artifactName` has been checked by `checkArtifactName` but `itemPath` not?